File: countingInversions/java/src/main/java/CountingInversions.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class CountingInversions {

    public static int countInversions(int[] input) {

        if(input == null || input.length <=1) {
            return 0;
        }

        CountInversionsResult countInversionsResult = counting(input);

        return countInversionsResult.count;
    }

    private static CountInversionsResult counting(int[] input) {
        if(input.length <=1) {
            CountInversionsResult countInversionsResult = new CountInversionsResult();
            countInversionsResult.count = 0;
            countInversionsResult.sortedInput = input;

            return countInversionsResult;
        }

        int mid = input.length/2;

        CountInversionsResult lCountInversionsResult = counting(Arrays.copyOfRange(input, 0, mid));
        int lCount = lCountInversionsResult.count;
        int[] linput = lCountInversionsResult.sortedInput;

        CountInversionsResult rCountInversionsResult = counting(Arrays.copyOfRange(input, mid, input.length));
        int rCount = rCountInversionsResult.count;
        int[] rinput = rCountInversionsResult.sortedInput;

        int count = 0;
        int idx = 0;
        int lidx = 0;
        int ridx = 0;
        int[] sortedArray = new int[input.length];

        while (lidx < linput.length || ridx < rinput.length) {
            if (lidx == linput.length) {
                sortedArray[idx] = rinput[ridx];
                ridx++;
            } else if (ridx == rinput.length) {
                sortedArray[idx] = linput[lidx];
                lidx++;
            } else if (linput[lidx] <= rinput[ridx]) {
                sortedArray[idx] = linput[lidx];
                lidx++;
            } else {
                sortedArray[idx] = rinput[ridx];
                ridx++;
                count += linput.length-lidx;
            }
            idx++;
        }

        CountInversionsResult countInversionsResult = new CountInversionsResult();
        countInversionsResult.sortedInput = sortedArray;
        countInversionsResult.count = count+lCount+rCount;
        return countInversionsResult;
    }

}

File: countingInversions/python3/countingInversions.py
def countInversions(input):
    if input == None or len(input) <=1:
        return 0

    [sorted_array, count] = counting(input)

    return count


def counting(input):
    if len(input) <= 1:
        return input, 0
    
    mid = len(input)//2
    [linput, lcount] = counting(input[0:mid])
    [rinput, rcount] = counting(input[mid:])

    count = 0
    lidx = 0
    ridx =0
    sort_array = []

    while lidx < len(linput) or ridx < len(rinput):
        if lidx == len(linput):
            sort_array.extend(rinput[ridx:])
            ridx = len(rinput)
        elif ridx == len(rinput):
            sort_array.extend(linput[lidx:])
            lidx = len(linput)
        elif linput[lidx] <= rinput[ridx]:
            sort_array.append(linput[lidx])
            lidx = lidx+1
        else:
            sort_array.append(rinput[ridx])
            ridx = ridx+1
            count = count + len(linput) - lidx
    
    return sort_array, count+lcount+rcount
